### PR TITLE
[Merged by Bors] - feat(Data/Sum): forall_sum_pi

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Sum/Basic.lean [hunk 1]
```diff
@@ -59,6 +59,20 @@ theorem «exists» {p : Sum α β → Prop} : (∃ x, p x) ↔ (∃ a, p (inl a)
     | Or.inr ⟨b, h⟩ => ⟨inr b, h⟩⟩
 #align sum.exists Sum.exists
 
+theorem forall_sum_pi (p : α ⊕ β → Sort*)
+    (q : (∀ a, p a) → Prop) :
+    (∀ a, q a) ↔ (∀ a b, q (Sum.rec a b)) :=
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Basic.lean:64]: Can you add the types of `a` and `b` here for clarity?

Comment (ChrisHughes24) [reply Mathlib/Data/Sum/Basic.lean]: done. I also made some arguments implict and improved some other names

File: Mathlib/Data/Sum/Basic.lean [hunk 2]
```diff
@@ -59,6 +59,18 @@ theorem «exists» {p : Sum α β → Prop} : (∃ x, p x) ↔ (∃ a, p (inl a)
     | Or.inr ⟨b, h⟩ => ⟨inr b, h⟩⟩
 #align sum.exists Sum.exists
 
+theorem forall_sum_pi {γ : α ⊕ β → Sort*} (p : (∀ ab, γ ab) → Prop) :
+    (∀ fab, p fab) ↔ (∀ fa fb, p (Sum.rec fa fb)) :=
+  ⟨fun h fa fb => h _, fun h fab => by
+    have h1 : fab = Sum.rec (fun i => fab (Sum.inl i)) (fun i => fab (Sum.inr i)) := by
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Basic.lean:65]: nit:
```suggestion
    have h1 : fab = Sum.rec (fun a => fab (Sum.inl a)) (fun b => fab (Sum.inr b)) := by
```

File: Mathlib/Data/Sum/Basic.lean [hunk 3]
```diff
@@ -59,6 +59,18 @@ theorem «exists» {p : Sum α β → Prop} : (∃ x, p x) ↔ (∃ a, p (inl a)
...
+      ext ab; cases ab <;> rfl
+    rw [h1]; exact h _ _⟩
+
+theorem exists_sum_pi {γ : α ⊕ β → Sort*} (p : (∀ ab, γ ab) → Prop) :
+    (∃ fab, p fab) ↔ (∃ fa fb, p (Sum.rec fa fb)) := by
+  rw [← not_forall_not, forall_sum_pi]
+  simp
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Basic.lean:72]: I'm curious if this brings in classical logic, though I don't really care.

**Review by eric-wieser — APPROVED** [review]
bors d+

These might be worth giving docstrings.

File: Mathlib/Data/Sum/Basic.lean (hunk 3 above)
Comment (ChrisHughes24) [reply Mathlib/Data/Sum/Basic.lean:72]: It does

File: Mathlib/Data/Sum/Basic.lean [hunk 4]
```diff
@@ -59,6 +59,18 @@ theorem «exists» {p : Sum α β → Prop} : (∃ x, p x) ↔ (∃ a, p (inl a)
     | Or.inr ⟨b, h⟩ => ⟨inr b, h⟩⟩
 #align sum.exists Sum.exists
 
+theorem forall_sum_pi {γ : α ⊕ β → Sort*} (p : (∀ ab, γ ab) → Prop) :
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Basic.lean:62]: I think we have a similar lemma for fin tuples; can you check if the naming is consistent?

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**